File: application/client/src/app/schema/render/dlt.ts
import { Render } from './index';
import { Columns } from './columns';
import { Protocol } from '@platform/types/observe/parser/index';

const MIN_COLUMN_WIDTH = 30;
const MAX_COLUMN_WIDTH = 600;
// public readonly widths: number[] = [150, 80, 80, 80, 80, 80, 80, 80, 80, 80, -1];

export class Implementation extends Render<Columns> {
    public static HEADERS = [
        {
            caption: 'Datetime',
            desc: 'Datetime',
        },
        {
            caption: 'ECUID',
            desc: 'ECU',
        },
        {
            caption: 'VERS',
            desc: 'Dlt Protocol Version (VERS)',
        },
        {
            caption: 'SID',
            desc: 'Session ID (SEID)',
        },
        {
            caption: 'MCNT',
            desc: 'Message counter (MCNT)',
        },
        {
            caption: 'TMS',
            desc: 'Timestamp (TMSP)',
        },
        {
            caption: 'EID',
            desc: 'ECU',
        },
        {
            caption: 'APID',
            desc: 'Application ID (APID)',
        },
        {
            caption: 'CTID',
            desc: 'Context ID (CTID)',
        },
        {
            caption: 'MSTP',
            desc: 'Message Type (MSTP)',
        },
        {
            caption: 'PAYLOAD',
            desc: 'Payload',
        },
    ];

    constructor() {
        super();
        this.setBoundEntity(
            new Columns(
                Implementation.HEADERS,
                true,
                [150, 20, 20, 20, 20, 20, 20, 20, 20, 20, -1],
                MIN_COLUMN_WIDTH,
                MAX_COLUMN_WIDTH,
            ),
        );
    }

    public override protocol(): Protocol {
        return Protocol.Dlt;
    }

    public override columns(): number {
        return Implementation.HEADERS.length;
    }
    public override delimiter(): string | undefined {
        return `\u0004`;
    }
}

File: application/apps/indexer/gui/application/src/session/ui/definitions/schema/dlt.rs
use std::ops::Range;

use egui_table::Column;
use parsers::dlt::fmt::DLT_COLUMN_SENTINAL;

use super::{ColumnInfo, LogSchema, map_columns_with_separator};

const MIN_COLUMN_WIDTH: f32 = 30.0;
const MAX_COLUMN_WIDTH: f32 = 600.0;

#[derive(Debug)]
pub struct DltLogSchema {
    columns: [ColumnInfo; 11],
}

impl Default for DltLogSchema {
    fn default() -> Self {
        let columns = [
            ColumnInfo::new("Datetime", "Datetime", dlt_column(150.0)),
            ColumnInfo::new("ECUID", "ECU", dlt_column(30.0)),
            ColumnInfo::new("VERS", "Dlt Protocol Version (VERS)", dlt_column(30.0)),
            ColumnInfo::new("SID", "Session ID (SEID)", dlt_column(30.0)),
            ColumnInfo::new("MCNT", "Message counter (MCNT)", dlt_column(30.0)),
            ColumnInfo::new("TMS", "Timestamp (TMSP)", dlt_column(30.0)),
            ColumnInfo::new("EID", "ECU", dlt_column(30.0)),
            ColumnInfo::new("APID", "Application ID (APID)", dlt_column(30.0)),
            ColumnInfo::new("CTID", "Context ID (CTID)", dlt_column(30.0)),
            ColumnInfo::new("MSTP", "Message Type (MSTP)", dlt_column(30.0)),
            ColumnInfo::new("PAYLOAD", "Payload", Column::default()),
        ];

        Self { columns }
    }
}

fn dlt_column(width: f32) -> Column {
    Column::new(width).range(MIN_COLUMN_WIDTH..=MAX_COLUMN_WIDTH)
}

impl LogSchema for DltLogSchema {
    fn has_headers(&self) -> bool {
        true
    }

    fn columns(&self) -> &[ColumnInfo] {
        &self.columns
    }

    fn map_columns(&self, log: &str) -> Vec<Range<usize>> {
        let mut ranges = Vec::with_capacity(self.columns.len());
        map_columns_with_separator(log, &mut ranges, DLT_COLUMN_SENTINAL);

        ranges
    }
}
